src/Graph.ts

```diff
@@ -454,6 +455,14 @@ export default class Graph {
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
 					visit(<Module>depModule, seen);
+					cycles.forEach(cycle => {
+						if (cycle.path[depModule.id] === null) {
+							cycle.path[depModule.id] = module.id;
+							if (cycle.id !== module.id) {
+								cycle.path[module.id] = null;
+							}
+						}
+					})
```

Comment:
I'd still like to see the complexity reduced here if possible. We are looping each cycle for each dependency for each module and doing a dictionary lookup. A graph like A -> (B,C,D), B -> A, C -> A, D -> ... where the D branch is deep will result in unnecessary looping over the former A,B,C cycles for each module of the D branch, which seems very unnecessary.

Reply:
Just to make sure we have the same understanding of this code: Your example will result in three dictionary lookups per module in the `D` branch, right?

Reply:
I can imagine marking the cycle as closed in L461 (by moving it out of `cycles` into some other array), that would remove unnecessary lookups in your example, but there are probably pathological cases were this wouldn't help at all.

Reply:
Yes. This is why I was thinking of parent-based seen as one could do:

```js
if (seen[id]) {
  let cycleStack = [];
  let parentId;
  while (parentId = seen[id]) {
    cycleStack.push(parentId);
    id = parentId;
  }
}
seen[id] = parentId; // with the parent added to the visitor to do this...
```

Each individual `seen` will only have one parent path per module by the nature of `seen`, guaranteeing the above stack which then only iterates the exact path. Although I may be overlooking some of the details.

Reply:
That would make sense, since `seen` is only used as a boolean anyway.

Reply:
Note that we may catch cases that aren't actually cycles - eg:

```
A -> B -> C
    -> C
```

C will be in seen, when it is hit again as a parallel dependency not a cyclical one. The way to confirm that it isn't cyclical is that when running up the parent path from the current C (direct dependency of A), we can note that we don't come back around to C. Whether as if we had:

```
C -> A -> B -> C
```

When we come across the second seen "C" and run up the parent stack from the current id, we do find the loop.

Reply:
Are you sure about this? It looks to me like modules are traversed depth-first.